### ISSUE TITLE
System.DivideByZeroException in the MovePart activity

### ISSUE THREAD

**Issue by abcdefg30** [OpenRA/OpenRA#11857]
No idea how I got that. I was just ordering an attack move while scrolling (actually, I think the game had a bug there and let my viewport jump to the bottom left edge, but I can't reproduce that reliably either).

```
Red Alert Mod at Version {DEV_VERSION}
on map 83e3d202a01e95b77fd776562aa418c7dca56a97 (04:  Ten to one by Westwood Studios).
Date: 2016-08-21 16:38:04Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.DivideByZeroException`: Attempted to divide by zero. (Es wurde versucht, durch 0 (null) zu teilen.)
   at OpenRA.Mods.Common.Activities.Move.MovePart..ctor(Move move, WPos from, WPos to, Int32 fromFacing, Int32 toFacing, Int32 startingFraction)
   at OpenRA.Mods.Common.Activities.Move.MoveFirstHalf..ctor(Move move, WPos from, WPos to, Int32 fromFacing, Int32 toFacing, Int32 startingFraction)
   at OpenRA.Mods.Common.Activities.Move.MoveFirstHalf.OnComplete(Actor self, Mobile mobile, Move parent)
   at OpenRA.Mods.Common.Activities.Move.MovePart.InnerTick(Actor self, Mobile mobile)
   at OpenRA.Mods.Common.Activities.Move.MovePart.Tick(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in \OpenRA\OpenRA.Game\Traits\ActivityUtils.cs:Line 37.
   at OpenRA.Mods.Common.Activities.AttackMoveActivity.Tick(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in \OpenRA\OpenRA.Game\Traits\ActivityUtils.cs:Line 37.
   at OpenRA.Actor.Tick() in \OpenRA\OpenRA.Game\Actor.cs:Line 164.
   at OpenRA.World.Tick() in \OpenRA\OpenRA.Game\World.cs:Line 339.
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in \OpenRA\OpenRA.Game\Game.cs:Line 561.
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:Line 586.
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:Line 713.
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:Line 753.
   at OpenRA.Program.Run(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 119.
   at OpenRA.Program.Main(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 40.
```

**Comment by abcdefg30** [comment]
I guess the problematic line is this one: https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/Move/Move.cs#L321

**Comment by RoosterDragon** [comment]
Anything attempting to change facing by 180 degrees in a single move action is going to cause the `u` and `v` vectors to be pointing in opposite directions, which would cause `v.X * u.Y - v.Y * u.X` to be 0. We should probably handle that.

Not sure how one can invoke that in gameplay though.

**Comment by cmoski** [comment]
Completely unhelpful I know, but I've encountered this crash randomly while on the most current head revision of OpenRA on the RA2 Mod located at https://github.com/OpenRA/ra2 -- looks like the dog entity triggers this movement somehow, although I didn't have time to sift through the code and figure out exactly how this was triggered.

Red Alert 2 Mod at Version {DEV_VERSION}
on map 2e19dd9bcba0c57d5cfc7d1f03c0ce06837685b7 (Powder Keg by Westwood Studios).
Date: 2016-08-24 08:30:15Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.DivideByZeroException`: Attempted to divide by zero.
   at OpenRA.Mods.Common.Activities.Move.MovePart..ctor(Move move, WPos from, WPos to, Int32 fromFacing, Int32 toFacing, Int32 startingFraction)
   at OpenRA.Mods.Common.Activities.Move.MoveFirstHalf..ctor(Move move, WPos from, WPos to, Int32 fromFacing, Int32 toFacing, Int32 startingFraction)
   at OpenRA.Mods.Common.Activities.Move.MoveFirstHalf.OnComplete(Actor self, Mobile mobile, Move parent)
   at OpenRA.Mods.Common.Activities.Move.MovePart.InnerTick(Actor self, Mobile mobile)
   at OpenRA.Mods.Common.Activities.Move.MovePart.Tick(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in C:\Dev\OpenRA-bleed\OpenRA.Game\Traits\ActivityUtils.cs:line 37
   at OpenRA.Mods.Common.Activities.AttackMoveActivity.Tick(Actor self)
   at OpenRA.Traits.ActivityUtils.RunActivity(Actor self, Activity act) in C:\Dev\OpenRA-bleed\OpenRA.Game\Traits\ActivityUtils.cs:line 37
   at OpenRA.Actor.Tick() in C:\Dev\OpenRA-bleed\OpenRA.Game\Actor.cs:line 163
   at OpenRA.World.Tick() in C:\Dev\OpenRA-bleed\OpenRA.Game\World.cs:line 338
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in C:\Dev\OpenRA-bleed\OpenRA.Game\Game.cs:line 561
   at OpenRA.Game.LogicTick() in C:\Dev\OpenRA-bleed\OpenRA.Game\Game.cs:line 586
   at OpenRA.Game.Loop() in C:\Dev\OpenRA-bleed\OpenRA.Game\Game.cs:line 713
   at OpenRA.Game.Run() in C:\Dev\OpenRA-bleed\OpenRA.Game\Game.cs:line 753
   at OpenRA.Program.Run(String[] args) in C:\Dev\OpenRA-bleed\OpenRA.Game\Support\Program.cs:line 119
   at OpenRA.Program.Main(String[] args) in C:\Dev\OpenRA-bleed\OpenRA.Game\Support\Program.cs:line 40

**Comment by pchote** [comment]
> Not sure how one can invoke that in gameplay though.

This shouldn't be possible.  I'm still waiting on @abcdefg30's replay so that we can debug this.

**Comment by GraionDilach** [comment]
This happens in 2 games out of 5 in AS. This is possible.

**Comment by pchote** [comment]
I was referring specifically to 180 degree turns without an intermediate `Turn` activity.

**Comment by GraionDilach** [comment]
Replay of this crash on https://github.com/OpenRA/OpenRA/commit/fab807b86ef4cc8dfad3b4b3f624e6aa24520f0a RA mod: https://drive.google.com/open?id=0B703_aY9zBYFeVZNbk8xYTdWcVE

**Comment by pchote** [comment]
You appear to have used modified game rules during that replay: the speed `superfast` doesn't exist.

**Comment by GraionDilach** [comment]
Ah, yeah, didn't expected replays storing speed - all I did was to have a gamespeed with a timestep of 10.

**Comment by cmoski** [comment]
After applying this fix, I found that in Move.cs MovePart...

                        var s = (v.Y * w.X - v.X * w.Y) * 1024 / (v.X * u.Y - v.Y * u.X);
                        var x = from.X + s * u.X / 1024;
                        var y = from.Y + s * u.Y / 1024;

(v.X * u.Y - v.Y * u.X) can actually still evaluate to 0 tossing a div by zero exception..

**Comment by pchote** [comment]
Please provide a replay and commit details.

**Comment by GraionDilach** [comment]
Requesting reopening:

Replay: https://drive.google.com/open?id=0B703_aY9zBYFVFdtcGJ4VlIxWTQ

TS mod, version https://github.com/OpenRA/OpenRA/tree/b7d0947d7dedefcffbb9c618b1325cca665fed20 on fastest gamespeed.

**Comment by pchote** [comment]
There happens because `Util.NormalizeFacing(-128)` is returning 127 instead of 128.  I haven't looked into why yet.

**Comment by Mailaender** [comment]
This is back:

![image](https://user-images.githubusercontent.com/756669/95662620-79889380-0b38-11eb-831c-de3983b35dfb.png)

See also https://github.com/Dzierzan/OpenSA/issues/131

**Comment by Mailaender** [comment]
Duplicate of https://github.com/OpenRA/OpenRA/issues/18325.
